src/hv_aic_events.c

```diff
@@ -0,0 +1,166 @@
+/* SPDX-License-Identifier: MIT */
+
+#include "adt.h"
+#include "hv.h"
+#include "uartproxy.h"
+
+#define AIC_EVENT      0x2004
+#define AIC_EVENT_TYPE GENMASK(31, 16)
+#define AIC_EVENT_NUM  GENMASK(15, 0)
+
+#define AIC_EVENT_TYPE_HW   1
+#define AIC_EVENT_TYPE_IPI  4
+#define AIC_EVENT_IPI_OTHER 1
+#define AIC_EVENT_IPI_SELF  2
+
+#define AIC_MAX_HW_NUM 1024
+
+static u64 aic_base;
+
+static u8 trace_hw_num[AIC_MAX_HW_NUM / 8];
+
+static void set_hw_num(u16 num)
+{
+    u8 *b = &trace_hw_num[num / 8];
+    *b |= 1 << (num & 0x7);
+}
+
+static void clear_hw_num(u16 num)
+{
+    u8 *b = &trace_hw_num[num / 8];
+    *b &= ~(1 << (num & 0x7));
+}
+
+static bool check_hw_num(u16 num)
+{
+    u8 val = trace_hw_num[num / 8];
+    return !!(val & (1 << (num & 0x7)));
+}
+
+static void emit_irqtrace(u16 type, u16 num)
+{
+    struct hv_evt_irqtrace evt = {
+        .flags = 0,
+        .type = type,
+        .num = num,
+    };
+
+    hv_wdt_suspend();
+    uartproxy_send_event(EVT_IRQTRACE, &evt, sizeof(evt));
+    hv_wdt_resume();
+}
+
+static bool trace_aic_event(u64 addr, u64 *val, bool write, int width)
+{
+    if (write) {
+        switch (width) {
+            case 0:
+                write8(addr, val[0]);
+                break;
+            case 1:
+                write16(addr, val[0]);
+                break;
+            case 2:
+                write32(addr, val[0]);
+                break;
+            case 3:
+                write64(addr, val[0]);
+                break;
+            case 4:
+                write64(addr, val[0]);
+                write64(addr + 8, val[1]);
+                break;
+            default:
+                dprintf("HV: unsupported width %ld\n", width);
+                return false;
+        }
+    } else {
+        switch (width) {
+            case 0:
+                val[0] = read8(addr);
+                break;
+            case 1:
+                val[0] = read16(addr);
+                break;
+            case 2:
+                val[0] = read32(addr);
+                break;
+            case 3:
+                val[0] = read64(addr);
+                break;
+            case 4:
+                val[0] = read64(addr);
+                val[1] = read64(addr + 8);
+                break;
+            default:
+                dprintf("HV: unsupported width %ld\n", width);
+                return false;
+        }
+    }
+
+    if (addr != (aic_base + AIC_EVENT) || write || width != 2) {
+        return true;
+    }
+
+    u16 type = (*val & AIC_EVENT_TYPE) >> 16;
+    u16 num = *val & AIC_EVENT_NUM;
+
+    switch (type) {
+        case AIC_EVENT_TYPE_HW:
+            if (check_hw_num(num)) {
+                emit_irqtrace(type, num);
+            }
+            break;
+        default:
+            // ignore
+            break;
+    }
+
+    return true;
+}
+
+void hv_trace_aic_events(u32 type, u32 num, u32 active)
```

Comment:
And then s/active/flags/, passing `IRQTRACE_IRQ` to enable tracing the IRQ firing.

Also, it might make sense to have a `u32 from, u32 count` instead of num, so that e.g. tracing all IRQs doesn't require hundreds of calls.

Let's also call this `hv_trace_irq` for consistency.